Halos/HaloModels/TNFW.py

```diff
@@ -0,0 +1,276 @@
+    r_trunc = tau * Rs
+    func = (r_trunc ** 2 * (-2 * x * (1 + r_trunc ** 2) + 4 * (1 + x) * r_trunc * np.arctan(x / r_trunc) -
+                            2 * (1 + x) * (-1 + r_trunc ** 2) * np.log(Rs) + 2 * (1 + x) * (-1 + r_trunc ** 2) * np.log(Rs * (1 + x)) +
+                            2 * (1 + x) * (-1 + r_trunc ** 2) * np.log(Rs * r_trunc) -
+                            (1 + x) * (-1 + r_trunc ** 2) * np.log(Rs ** 2 * (x ** 2 + r_trunc ** 2)))) / (2. * (1 + x) * (1 + r_trunc ** 2) ** 2)
+    mass_loss = func / (np.log(1+c)-c/(1+c))
+    return mass_loss
+    
+def tau_mf_interpolator():
+
+    N = 250
+    tau = np.logspace(-3.5, 2.5, N)
+    concentration = np.linspace(1.0, 200.0, N)
+    
+    log10_mass_fraction_1d = np.linspace(-4, -0.001, N)
+    log10tau_2d = np.zeros((N, N))
+
+    # This computes the value of tau that correponds to each pair of (concentration, mass_loss) 
+    for i, con_i in enumerate(concentration):
+        mfinal = tnfw_mass_fraction(tau, con_i)
+        log10final_mass = np.log10(mfinal)
+        mfinterp = interp1d(log10final_mass, np.log10(tau), fill_value='extrapolate')
+        
+        for j, log10_mass_j in enumerate(log10_mass_fraction_1d):
+            log10tau_2d[i,j]  = float(mfinterp(log10_mass_j))
+
+    interp_points = (concentration, log10_mass_fraction_1d)
+    interpolator = RegularGridInterpolator(interp_points, log10tau_2d, fill_value=None, bounds_error=False)
+    return interpolator
+
+truncation_radius_interpolator = tau_mf_interpolator()
+
+class TNFWSubhaloEmulator(Halo):
+    """
+    Defines a truncated NFW halo that is a subhalo of the host dark matter halo
+    """
+    def __init__(self, infall_mass, x, y, final_bound_mass, infall_concentration, redshift,
+                 lens_cosmo_instance, unique_tag=None):
+        """
+        See documentation in base class (Halos/halo_base.py)
+        """
+        self._lens_cosmo = lens_cosmo_instance
+
+        r3d = None
+        profile_definition = 'TNFW'
+        sub_flag = True
+        args = None
+        if unique_tag is None:
+            unique_tag = np.random.rand()
+
+        # set the concentration
+        self.c = infall_concentration
+        self._bound_mass_fraction = final_bound_mass/infall_mass
+        x_arcsec = x / self._lens_cosmo.cosmo.kpc_proper_per_asec(redshift)
+        y_arcsec = y / self._lens_cosmo.cosmo.kpc_proper_per_asec(redshift)
+        super(TNFWSubhaloEmulator, self).__init__(infall_mass, x_arcsec, y_arcsec, r3d, profile_definition, redshift, sub_flag,
+                                           lens_cosmo_instance, args, unique_tag)
+
+    @property
+    def lenstronomy_params(self):
+        """
+        See documentation in base class (Halos/halo_base.py)
+        """
+        if not hasattr(self, '_kwargs_lenstronomy'):
+
+            [concentration, rt_kpc] = self.profile_args
+            Rs_angle, theta_Rs = self._lens_cosmo.nfw_physical2angle(self.mass, concentration, self.z)
```

Comment:
similar to the kpc_per_asec comment, this method gets called multiple times with the same inputs. It would be good to create a private function inside the class that evaluates Rs_angle and theta_Rs once. You can use the if not hasattr(self, ...) syntax to do this, just it is done in this function.